File: Python3.x/110-Balanced_Binary_Tree.py
# Runtime: 52 ms, faster than 94.10% of Python3 online submissions for Balanced Binary Tree.
# Memory Usage: 17.8 MB, less than 89.13% of Python3 online submissions for Balanced Binary Tree.
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        self.flag = True
        self.get_height(root)
        return self.flag

    def get_height(self, node):
        if not node:
            return 0
        left = self.get_height(node.left)
        right = self.get_height(node.right)
        if abs(left - right) > 1:
            self.flag = False
        return max(left, right) + 1

File: Algorithms/C++/110-Balanced_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return getHeight(root) >= 0;
    }

    int getHeight(TreeNode* root) {
        if (root == nullptr) return 0;
        int leftHeight = getHeight(root->left),
            rightHeight = getHeight(root->right);
        if (leftHeight >= 0 && rightHeight >= 0 && std::abs(leftHeight - rightHeight) <= 1) return std::max(leftHeight, rightHeight) + 1;
        return -1;
    }
};
